BaseClasses.py

```diff
@@ -334,11 +334,17 @@ def get_placeable_locations(self, state=None, player=None) -> List[Location]:
         return [location for location in self.get_locations() if
                 (player is None or location.player == player) and location.item is None and location.can_reach(state)]
 
-    def get_unfilled_locations_for_players(self, location_name: str, players: Iterable[int]):
+    def get_unfilled_locations_for_players(self, locations, players: Iterable[int]):
         for player in players:
-            location = self.get_location(location_name, player)
-            if location.item is None:
-                yield location
+            if len(locations) == 0:
+                locations = [item.name for item in self.get_unfilled_locations(player)]
+            for location_name in locations:
+                if (location_name, player) in self._location_cache:
```

Comment:
could be made slightly faster, since you check the location cache anyway. E.g.: `location = self._location_cache.get((location_name, player), None)